Add socketcan_bridge and can_msgs.

iwanders opened the pull request:
Currently there is no package available in ROS to expose can frames from SocketCAN to a topic or the other way around. This pull request contains the additions which provide this functionality, the development has been discussed in issue #137.

It adds the `can_msgs` package, which holds a message type for a CAN Frame. The `socketcan_bridge` package contains nodes and classes that can bridge frames from SocketCAN to a ROS Topic and vice versa. 

The `socketcan_bridge` builds on the `socketcan_interface` package and does not depend on other `ros_canopen` packages. One change to `socketcan_interface` was required: the `DummyInterface` has been modified in order to facilitate testing the `socketcan_bridge` functionality, test results are unaffected by the changes.

Please review the code, if there are any obstacles to the merge, please let me know. I'll write the wiki pages and send an announcement on the ros-users list when the time comes.

iwanders committed:
Introduces the can_msgs package for message types.

Package to hold CAN message types, the Frame message type can contain the data
as returned by SocketCAN.

iwanders committed:
Introduces socketcan_to_topic and its node.

This is based on the ThreadedSocketCANInterface from the socketcan_interface package. Sending might become problematic with this class however, as the send() method is not exposed through the Threading wrappers.

iwanders committed:
Introduces topic_to_socketcan and the bridge.

Adds TopicToSocketCAN, the counterpart to the SocketCANToTopic class.
Also introduces a node to use this class and a node which combines the two
classes into a bidirectional bridge.

Contrary to the previous commit message the send() method appears to be
available from the ThreadedSocketCANInterface afterall.

iwanders committed:
Adds conversion test between msg and SocketCAN Frame.

This test only covers the conversion between the can_msgs::Frame message and can::Frame from SocketCAN.

iwanders committed:
Changes inheritance of DummyInterface to DriverInterface.

Such that this interface can also be used for tests requiring a DriverInterface
class.

Test results of the socketcan_interface tests are unchanged by this
modification as it only uses the CommInterface methods.

iwanders committed:
Adds tests for socketcan_bridge and bugfixes.

Uses rostests and the modified DummyInterface to test whether behaviour
is correct. Prevented possible crashes when can::tostring was called on
invalid frames.

iwanders committed:
Finalizes work on the socketcan_bridge and can_msgs.

Readies the packages socketcan_bridge and can_msgs for the merge with ros_canopen.
Bumps the version for both packages to 0.1.0. Final cleanup in CMakeLists, added
comments to the shell script and launchfile used for testing.

mathias-luedtke commented:
Thank you very much for your contribution!

I have just three points:
1. `can_msgs` lacks `message_runtime`, please update it with the help of this guide: http://docs.ros.org/indigo/api/catkin/html/howto/format2/building_msgs.html
~~And you do not use `std_msgs`, so please remove the dependency.~~ (Header will be replaced by std_msgs/Header internally)
2. Where did you get the `header.frame_id = "0"` from, I am not aware of this convention.
3. It would be great if `DummyInterface` would actually implement the StateInterface to a certain extent. At least `doesLoopBack()` should reflect the loopback option.

mathias-luedtke commented on socketcan_bridge/CMakeLists.txt:
```diff
@@ -0,0 +1,75 @@
+
+catkin_package(
+  INCLUDE_DIRS include
+  LIBRARIES socketcan_to_topic topic_to_socketcan socketcan_interface roscpp can_msgs
```
`socketcan_interface roscpp can_msgs` should go into `CATKIN_DEPENDS`

mathias-luedtke commented on socketcan_bridge/CMakeLists.txt:
```diff
@@ -0,0 +1,75 @@
+
+add_library(socketcan_to_topic src/socketcan_to_topic.cpp)
+target_link_libraries(socketcan_to_topic ${catkin_LIBRARIES} ${Boost_LIBRARIES})
+add_dependencies(socketcan_to_topic can_msgs_generate_messages_cpp)
```
The recommended way is to depend on `${catkin_EXPORTED_TARGETS}`, see http://docs.ros.org/indigo/api/catkin/html/howto/format2/cpp_msg_dependencies.html

mathias-luedtke commented on socketcan_bridge/CMakeLists.txt:
```diff
@@ -0,0 +1,75 @@
+
+add_library(topic_to_socketcan src/topic_to_socketcan.cpp)
+target_link_libraries(topic_to_socketcan ${catkin_LIBRARIES} ${Boost_LIBRARIES})
+add_dependencies(topic_to_socketcan can_msgs_generate_messages_cpp)
```
same here

mathias-luedtke commented on socketcan_bridge/CMakeLists.txt:
```diff
@@ -0,0 +1,75 @@
+
+install(DIRECTORY include/${PROJECT_NAME}/
+  DESTINATION ${CATKIN_PACKAGE_INCLUDE_DESTINATION}
+  FILES_MATCHING PATTERN "*.h")
```
I had troubles in the past with the `FILES_MATCHING` directive. It is not needed, better remove it.

mathias-luedtke commented on socketcan_bridge/CMakeLists.txt:
```diff
@@ -0,0 +1,75 @@
+  DESTINATION ${CATKIN_PACKAGE_INCLUDE_DESTINATION}
+  FILES_MATCHING PATTERN "*.h")
+
+roslint_cpp()
```
roslint is a test_depend, so maybe the should be moved into the `if(CATKIN_ENABLE_TESTING)` block?
(Including the find package directive)

mathias-luedtke commented on socketcan_bridge/src/socketcan_to_topic.cpp:
```diff
@@ -0,0 +1,96 @@
+
+      state_listener_ = driver_->createStateListener(
+              can::StateInterface::StateDelegate(this, &SocketCANToTopic::stateCallback));
+    };
```
return is missing.
What should be return? I would stick to bool (false means error) or change it to void.

mathias-luedtke commented on socketcan_bridge/src/topic_to_socketcan.cpp:
```diff
@@ -0,0 +1,91 @@
+    {
+      state_listener_ = driver_->createStateListener(
+              can::StateInterface::StateDelegate(this, &TopicToSocketCAN::stateCallback));
+    };
```
return is missing.

mathias-luedtke commented on socketcan_bridge/src/topic_to_socketcan_node.cpp:
```diff
@@ -0,0 +1,66 @@
+  if (!driver->init(can_device, 0))  // initialize device at can_device, 0 for no loopback.
+  {
+    ROS_FATAL("Failed to initialize can_device at %s", can_device.c_str());
+    return -1;
```
please use exit code 1 for failures. For some reason canopen_chain_node and canopen_motor_node use -1, but that is wrong and should be fixed.

iwanders committed:
Adds message_runtime to can_msgs dependencies.

Added the missing dependency, also changes message_generation to a build_depend.

iwanders committed:
Cleanup, fixes and improvements in CmakeLists.

Adds the REQUIRED COMPONENTS packages to the CATKIN_DEPENDS.
Improves add_dependency on the messages to be built.
Removes unnecessary FILES_MATCHING.
Moves the roslint_cpp macro to the testing block.

gavanderhoorn commented:
@ipa-mdl wrote:

> 2 . Where did you get the `header.frame_id = "0"` from, I am not aware of this convention.

I'm not sure this is actually still valid / best practice, but it is documented in [std_msgs/Header](http://docs.ros.org/jade/api/std_msgs/html/msg/Header.html):

```
#Frame this data is associated with
# 0: no frame
# 1: global frame
string frame_id
```

mathias-luedtke commented:
@gavanderhoorn: thanks for the hint, I knew I have seen it somewhere before..

I would suggest to leave the frame empty. I think this is the convention for the joint_states topic.

gavanderhoorn commented:
Yes, I wouldn't put a `0` or a `1` in there either. Might actually be something to ask on ROS Answers / `ros-users`.

mathias-luedtke commented:
Related answer by @adolfo-rt: http://answers.ros.org/question/74422/ros-gazebo-plugin-joint-state-topic-frame_id-is-blank/
A CAN bus does not have a single location, so empty should be fine.

iwanders committed:
Changes return type of setup() method.

Setup() calls the CreateMsgListener and CreateStateListener, it does not attempt
to verify if the interface is ready, which makes void more applicable.

iwanders committed:
Changes the frame field for the published messages.

An empty frame name is more commonly used to denote that there is no frame
associated with the message.

iwanders committed:
Changes the exit code of the nodes if device init fails.

Now exits with 1 if the initialization of the can device fails.

iwanders commented:
Thanks for the extensive review. I'm new to catkin and CMake, the links were much appreciated.

> I had troubles in the past with the FILES_MATCHING directive

That exact line is also produced by calling `catkin_create_pkg`, (although I copied it from `canopen_master`). I read the CMake [install](https://cmake.org/cmake/help/v3.0/command/install.html) documentation; it only serves as a filter. Since there are only .h files in this directory we can do without it indeed.

The header `"0"` was indeed from that the page @gavanderhoorn mentioned. I have changed it to the empty frame as proposed.

I changed the return type of the `setup()` method to void, the `CreateMsgListener` does not provide any indication whether it completed successfully. We could of course check the state of the provided `driver`, but that state is already checked when the `driver::init` when the node is started. To me, it would make no sense to perform these checks twice.

I updated the exit codes, these were indeed copied from the `canopen_chain_node`.

>  It would be great if `DummyInterface` would actually implement the `StateInterface` to a certain extent.

I totally agree, that would also provide a way to test whether the `ROS_INFO` and `ROS_ERROR` messages from the `stateCallback` are working. Is there any standard way for automated testing whether such messages were sent? Or is it acceptable to verify this by hand?

mathias-luedtke commented:
> That exact line is also produced by calling catkin_create_pkg, (although I copied it from canopen_master).

You are right, the problem was not `FILES_MATCHING` , but the combination with `PATTERN ".svn" EXCLUDE` (as in the catkin_create_pkg template). Sorry for the noise.

> CreateMsgListener does not provide any indication whether it completed successfully.

The current implementation might throw if the system runs out of memory.
You could check that the shared pointer is valid, but this case cannot happen for now.

> Is there any standard way for automated testing whether such messages were sent? Or is it acceptable to verify this by hand?

I don't think that there is a standard way for automated test. You could listen to rosout and check it, but that is not necessary IMHO.

However, you cannot rely on rosout for user to get notified of bus errors, the log output can be muted. Perhaps it would make sense to add diagnostics (http://wiki.ros.org/diagnostic_updater) , but this is really just nice to have.

gavanderhoorn commented:
@ipa-mdl wrote:

> @iwanders wrote:
> 
> > That exact line is also produced by calling catkin_create_pkg, (although I copied it from canopen_master).
> 
> You are right, the problem was not FILES_MATCHING , but the combination with PATTERN ".svn" EXCLUDE (as in the catkin_create_pkg template). Sorry for the noise.

That filter was really only necessary with old svn versions, as they used to create `.svn` subdirs in all directories of your project. As this is a git project, it is really not needed.

I'd not even bother with it and just remove it.

mathias-luedtke commented on socketcan_bridge/CMakeLists.txt:
```diff
@@ -0,0 +1,71 @@
+
+catkin_package(
+  INCLUDE_DIRS include
+  LIBRARIES socketcan_to_topic topic_to_socketcan socketcan_interface roscpp can_msgs
```
Please remove `socketcan_interface roscpp can_msgs`, `LIBRARIES` is meant for the library targets of this package.
Other libraries from the dependencies will be added automatically by catkin.

mathias-luedtke commented on socketcan_bridge/CMakeLists.txt:
```diff
@@ -0,0 +1,71 @@
+  INCLUDE_DIRS include
+  LIBRARIES socketcan_to_topic topic_to_socketcan socketcan_interface roscpp can_msgs
+  CATKIN_DEPENDS can_msgs roscpp socketcan_interface
+  DEPENDS Boost
```
No need to depend on Boost here, since you do not use it directly in your code.
You could keep it, but then you should add the dependency in package.xml as well.

mathias-luedtke commented:
The tests complete successfully now (they crashed because of the missing return).
And cangen on vcan works as well :)

One last issue:
gtest tests needs this: http://docs.ros.org/indigo/api/catkin/html/howto/format2/gtest_configuration.html

I have found this issue with the help of https://github.com/fkie/catkin_lint

iwanders committed:
Improves StateInterface implementation of the DummyInterface.

The doesLoopBack() method now returns the correct value. A state change is
correctly dispatched when the init() method is called.

iwanders committed:
Removes dependency on Boost, adds rosunit dependency.

The dependency on Boost was unnecessary, rosunit is required for gtest.

iwanders commented:
> No need to depend on Boost here, since you do not use it directly in your code.
> You could keep it, but then you should add the dependency in package.xml as well.

It was already in present in the package.xml, but I'm all for removing unnecessary things. I read some more documentation and I believe it is better to leave it out; the additions to the linking were also unnecessary.

> The tests complete successfully now (they crashed because of the missing return).

Excellent, which tests / tool did crash because of the missing return statements? I did not observe any errors from `catkin_make run_tests`, did you use an external tool to perform a static analysis?

> I have found this issue with the help of [https://github.com/fkie/catkin_lint](https://github.com/fkie/catkin_lint)

Thanks, another useful tool; it reports one more issue in the `socketcan_bridge`:

> socketcan_bridge: error: unknown test_depend 'gtest'

But this `test_depend` on `gtest` is actually from the wiki Catkin tutorial page; [line 37 of this example](http://wiki.ros.org/ROS/Tutorials/catkin/CreatingPackage#dependencies_tags). The tests run fine without the `<test_depend>gtest</test_depend>` in the package, I presume you agree with removing it?

mathias-luedtke commented:
> Excellent, which tests / tool did crash because of the missing return statements? I did not observe any errors from catkin_make run_tests, did you use an external tool to perform a static analysis?

Both add_rostest_gtest targets crashed and rostest just reported "no test output". It took me a while to investigate this. I am using clang++ per default and it seems to handle missing return different from gcc. However, it issued warnings ;)

> But this test_depend on gtest is actually from the wiki Catkin tutorial page; line 37 of this example.

I don't trust the wiki on catkin-related topics anymore, just stick to http://docs.ros.org/indigo/api/catkin/html/howto/format2/gtest_configuration.html and the other useful howtos.

> I presume you agree with removing it?

Yes :)

iwanders committed:
Removes gtest from test dependencies.

This dependency is covered by the rosunit dependency.

iwanders commented:
> I am using clang++ per default and it seems to handle missing return different from gcc. 

Ok, I'll look into using Clang in combination with Catkin.

> I don't trust the wiki on catkin-related topics anymore, just stick to [http://docs.ros.org/indigo/api/catkin/html/](http://docs.ros.org/indigo/api/catkin/html/)

Those pages are indeed a lot more comprehensive, I'll use those as reference from now on.

Thanks for the tips :)

mathias-luedtke commented:
Everything passes now, great work! Thanks again!
If you have some time left, it would be great if you could add some documentation in the wiki: http://wiki.ros.org/socketcan_bridge.

iwanders commented:
Great! 

> If you have some time left, it would be great if you could add some documentation in the wiki.

Yes, I'll do that once the spamming issue is resolved on the wiki and account creation is enabled again. I'll also try to get an announcement posted on the users list once the wiki is updated.
